lib/sugarcrm.js

```diff
@@ -33,314 +31,316 @@ const defaultPlatform = ELASTICIO_TASK_ID && ELASTICIO_STEP_ID
  * @constructor
  */
 module.exports = function Service(config, emitter) {
-    if (!config || !config.baseUrl) {
-        throw new Error('SugarCRM Configuration is missing base URL');
+  if (!config || !config.baseUrl) {
+    throw new Error('SugarCRM Configuration is missing base URL');
+  }
+
+  const cfgUrl = {};
+  const urlParts = urlParser.parse(config.baseUrl);
+
+  if (urlParts.protocol) {
+    cfgUrl.protocol = urlParts.protocol;
+  }
+
+  if (urlParts.hostname) {
+    cfgUrl.baseUrl = urlParts.hostname + urlParts.pathname;
+  } else {
+    cfgUrl.baseUrl = urlParts.pathname;
+  }
+
+  const cfg = { protocol: 'https:', ...cfgUrl, ...config };
+
+  cfg.platform = cfg.platform || defaultPlatform;
+
+  function buildUrl(urlPath) {
+    return `${cfg.protocol}//${path.join(cfg.baseUrl, API_URI, urlPath)}`;
+  }
+
+  /**
+   * Check the response from SugarCRM.  Returns true for 200s responses authenticated.
+   * Returns false if there is a problem with authentication.
+   * Throws an exception for all other response codes.
+   * @returns {boolean}
+   */
+  function wasAuthenticationSuccessful(sugarCrmRepsonse) {
+    if (sugarCrmRepsonse.statusCode >= 200 && sugarCrmRepsonse.statusCode < 300) {
+      return true;
     }
-    let cfg = {
-        protocol: 'https:'
-    };
-
-    const url = urlParser.parse(config.baseUrl);
 
-    if (url.protocol) {
-        config.protocol = url.protocol;
+    if (sugarCrmRepsonse.statusCode === 401 && sugarCrmRepsonse.body
+      && sugarCrmRepsonse.body.error === 'invalid_grant') {
```

Comment:
Why do you need to check a combination of next args: `statusCode === 401` && `sugarCrmRepsonse.body` && `sugarCrmRepsonse.body.error === 'invalid_grant'`?
Is it sufficient to return false when you eval next if-block as `false`?
```
    if (sugarCrmRepsonse.statusCode >= 200 && sugarCrmRepsonse.statusCode < 300) {
      return true;
    }
    else {
      return false;
    }
```

Reply:
We didn't touch this part because it works.
But I think author wanted to handle three cases in this function.
1. Auth OK = true
2. 401, Auth not OK = false
3. Unexpected response from server = exception